src/listeners.ts

```diff
@@ -0,0 +1,582 @@
+): boolean {
+    const shift = keydown_event.shiftKey;
+    const ctrl = keydown_event.ctrlKey || keydown_event.metaKey;
+    const key_is_z = (
+        keydown_event.key === "z" ||
+        keydown_event.key === "Z" ||
+        keydown_event.code === "KeyZ"
+    );
+
+    if (ctrl && key_is_z) {
+        keydown_event.preventDefault();
+        if (shift) {
+            ulabel.redo();
+        } else {
+            ulabel.undo();
+        }
+        return false;
+    } else {
+        const current_subtask = ulabel.get_current_subtask();
+        switch (keydown_event.key) {
+            case "Escape":
+                // If in erase or brush mode, cancel the brush
+                if (current_subtask.state.is_in_erase_mode) {
+                    ulabel.toggle_erase_mode();
+                } else if (current_subtask.state.is_in_brush_mode) {
+                    ulabel.toggle_brush_mode();
+                } else if (current_subtask.state.starting_complex_polygon) {
+                    // If starting a complex polygon, undo
+                    ulabel.undo();
+                } else if (current_subtask.state.is_in_progress) {
+                    // If in the middle of drawing an annotation, cancel the annotation
+                    ulabel.cancel_annotation();
+                }
+                break;
+        }
+    }
+}
+
+/**
+ * Create listeners for a ULabel instance.
+ * Consider breaking out long handlers.
+ *
+ * @param ulabel ULabel instance
+ */
+export function create_ulabel_listeners(
+    ulabel: ULabel,
+) {
+    // ================= Mouse Events in the ID Dialog =================
+    const id_dialog = $(".id_dialog");
+    id_dialog.on(
+        "mousemove" + ULABEL_NAMESPACE,
+        function (mouse_event) {
+            if (!ulabel.get_current_subtask()["state"]["idd_thumbnail"]) {
+                ulabel.handle_id_dialog_hover(mouse_event);
+            }
+        },
+    );
+
+    // ================= Mouse Events in the Annotation Container =================
+    const annbox = $("#" + ulabel.config["annbox_id"]);
+
+    // Detect and record mousedown
+    annbox.on(
+        "mousedown" + ULABEL_NAMESPACE,
+        function (click_event) {
+            ulabel.handle_mouse_down(click_event);
+        },
+    );
+
+    // Prevent default for auxclick
+    $(document).on(
+        "auxclick" + ULABEL_NAMESPACE,
+        ulabel.handle_aux_click,
+    );
+
+    // Detect and record mouseup
+    $(document).on(
+        "mouseup" + ULABEL_NAMESPACE,
+        ulabel.handle_mouse_up.bind(ulabel),
+    );
+
+    $(window).on(
+        "click" + ULABEL_NAMESPACE,
+        function (click_event) {
+            if (click_event.shiftKey) {
+                click_event.preventDefault();
+            }
+        },
+    );
+
+    // Mouse movement has meaning in certain cases
+    annbox.on(
+        "mousemove" + ULABEL_NAMESPACE,
+        function (move_event) {
+            ulabel.handle_mouse_move(move_event);
+        },
```

Comment:
As I was moving this, I couldn't help but wonder why some of these aren't reduced. Couldn't this just be `ulabel.handle_mouse_move` ?

Reply:
probably, i think if the handler actually needs access to the ulabel class then you'll need to bind the function to ulabel explicitly, something about how the reference to the function is actually mounted to the listener.

Reply:
Indeed, it seems that passing `ulabel.handle_mouse_move` calls the function with `this` referencing the "annbox", but `ulabel.handle_mouse_move.bind(ulabel)` seems to work just fine. 

I don't think there's too much readability change in doing this (maybe a little bit of a loss if someone isn't familiar with `.bind`), and it is does keep things concise. Do you think I should switch to this when possible?

Reply:
Alternative that doesn't increase line count is arrow functions, 
e.g. `move_event => ulabel.handle_mouse_move(move_event)`

Reply:
yeah both the `bind` option or `=>` functions seem nice, i like either one better than the `function` declaration

Reply:
I'll change all one-liners to `=>` functions. Would you prefer I make everything `=>` to match? I would likely keep indentation the same

Reply:
Also it appears the the linting rule for arrow functions prefers no parentheses for one arg, e.g. `arg => { ... }` instead of `(arg) => { ... }`. 

Relevant rule is [arrow-parens](https://eslint.style/rules/default/arrow-parens), I can change this if we have a preference.

Reply:
interesting, yeah i think it might be nice to be consistent and always enforce (arg)

Reply:
Changed the rule and made all inline handlers arrow functions in da9b151.